[B+C] Add Merchant Trade API - Adds BUKKIT-5663

gabizou opened the pull request:
**The Issue:**

There is no way to modify Villager trades through the Bukkit API.

**Justification**

Without the API to allow for modifications being made on a Merchant's offers list, plugins have no Bukkit API way to affect trades, leading to messy uses of NMS classes and modified CraftBukkit classes, which can cause plugins to break or the server to crash. With an API, all plugin developers will be able to safely modify merchant trades (which as of current writing is only implemented through Villagers).

**PR Breakdown:**

The API adds the ability to retrieve current trades offered by a Merchant, add trades, remove trades and modify trades. There is also the added ability to influence event specific trades such that modified trades added to the MerchantTradeEvent do not alter the more permanent TradeOffers supplied by the Merchant.
In this sense: A MerchantTradeEvent's offers are a copy of the Merchant's offers that do not affect the Merchant's offers. However, the Merchant is still able to have their trades altered, but the MerchantTradeEvent will not be affected.

**Testing Results / Materials:**

Plugin code used to test: https://github.com/gabizou/TradeAPITest
Compiled Binary: https://dl.dropboxusercontent.com/u/9058563/TradingTest.jar

Steps to test:
- Spawn villager
- Interact with villager to bring up the Trade window

Console shows the debugging text

**Associated PRs**
Bukkit PR: https://github.com/Bukkit/Bukkit/pull/1077
CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/1393 

*\* Relevant PRs*\* 
Closed Bukkit PR: https://github.com/Bukkit/Bukkit/pull/921
Closed Bukkit PR: https://github.com/Bukkit/Bukkit/pull/833
Closed Bukkit PR: https://github.com/Bukkit/Bukkit/pull/832
Closed CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/1240
Closed CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/1111

**JIRA Tickets**

Main ticket: https://bukkit.atlassian.net/browse/BUKKIT-5663
Subsequent (duplicate) tickets
https://bukkit.atlassian.net/browse/BUKKIT-3935
https://bukkit.atlassian.net/browse/BUKKIT-2138

mbax commented on src/main/java/org/bukkit/event/merchant/MerchantAddOfferEvent.java:
```diff
@@ -0,0 +1,78 @@
+
+    /**
+     * Returns a copy of the list of offers being modified. Modifying this
+     * offers list will not change the offers resulting in the linked
```
It can't be modified. Don't suggest it :P

mbax commented on src/main/java/org/bukkit/event/merchant/MerchantAddOfferEvent.java:
```diff
@@ -0,0 +1,78 @@
+        return handlers;
+    }
+
+    public HandlerList getHandlerList() {
```
static

mbax commented on src/main/java/org/bukkit/event/merchant/MerchantRemoveOfferEvent.java:
```diff
@@ -0,0 +1,77 @@
+        return handlers;
+    }
+
+    public HandlerList getHandlerList() {
```
static

mbax commented on src/main/java/org/bukkit/event/merchant/MerchantTradeEvent.java:
```diff
@@ -0,0 +1,117 @@
+     * @param offer to remove
+     */
+    public void removeOffer(TradeOffer offer) {
+        if (offer != null) {
```
Why nullchecking like this instead of Validate.notNull ?

gabizou replied:
Because Validate.notNull will throw an exception last I remember, I want to avoid throwing exceptions on things as simple as this.

Wolvereness replied:
That's ridiculous. If they pass null, they made a mistake, and they should get an exception in their face, just like much more of the API.

Wolvereness commented on src/main/java/org/bukkit/entity/Merchant.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public boolean removeOffer(TradeOffer offer);
+
+}
```
```
>.>
```

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,266 @@
+         */
+        PLUGIN_ADDED
+    }
+}
```
```
>.>
```

Wolvereness commented on src/main/java/org/bukkit/event/merchant/MerchantTradeEvent.java:
```diff
@@ -0,0 +1,117 @@
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
```
```
>.>
```

Wolvereness commented on src/main/java/org/bukkit/event/merchant/MerchantRemoveOfferEvent.java:
```diff
@@ -0,0 +1,77 @@
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+}
```
```
>.>
```

Wolvereness commented on src/main/java/org/bukkit/event/merchant/MerchantEvent.java:
```diff
@@ -0,0 +1,26 @@
+        return this.merchant;
+    }
+
+}
```
```
>.>
```

Wolvereness commented on src/main/java/org/bukkit/event/merchant/MerchantAddOfferEvent.java:
```diff
@@ -0,0 +1,78 @@
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+}
```
```
>.>
```

Wolvereness commented:
You leak mutable objects everywhere, which is unacceptable. The inventory API is bad enough as it is, we don't want to make it worse. Examine how ItemMeta works for a good example of how mutability needs to be handled; ItemStack always returns a copy. Any of the new events with items return clones.

gabizou commented:
> You leak mutable objects everywhere, which is unacceptable.
> Any of the new events with items return clones.

Should be fixed in latest commit. I've added a utility method to MerchantEvent that performs a deep copy of a list of TradeOffers through ImmutableList and added the relevant ItemStack.clone where necessary.

I should note, I initially didn't have such cloning immutability because of the pre-existing API leading me to believe that mutability was more of a standard and over immutability/cloning from getters would be counter productive. I appreciate knowing now that future API should have an general sense of immutability than mutability.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,268 @@
+        /**
+         * Represents a plugin generated trade offer.
+         */
+        PLUGIN_ADDED
```
Adding a comma at the end would be appropriate.

Wolvereness commented:
- `@Utility` was designed for `ItemStack` or other API (read: CB) where extending it is _intended_.
- `MerchantAddOfferEvent.getCurrentOffers()` leaks mutable references.
- Why isn't `TradeOffer` immutable to begin with?
- Can `TradeOffer` have all `public`/`protected` constructors removed? Something like fireworks or `Color`. Relying on parameter types without a good name isn't the best idea for a sane API, especially when I don't think extending it is intended.
- When accepting an `ItemStack` via API that goes back into CB, always use `new ItemStack(ItemStack)`, as we can't confirm the plugin doesn't have a malicious `clone()` method. `ItemMeta` instances are fine as long as they have been passed through the `ItemFactory` at some point (for example, in book events, it's compared to null, which throws an exception for malicious input).
- `TradeOffer` implements `equals(Object)` but not `hashCode()`.

I'm sure there are many more issues, but I can't take much more at the moment.

gabizou commented:
> Why isn't TradeOffer immutable to begin with?

It's possible, just, as I've said before, I was basing this off of the current Inventory API, I can refactor all of this to use a TradeOffer Builder just as well.

> Can TradeOffer have all public/protected constructors removed? Something like fireworks or Color. Relying on parameter types without a good name isn't the best idea for a sane API, especially when I don't think extending it is intended.

Yes. I'll refactor to a static builder to produce an immutable TradeOffer instead of constructors.

gabizou commented:
My one worry is that doing this breaks a line where TradeOfferMeta isn't used for a singular Item or Entity, but can have several. Is it your intention to see this as an example:

```
TradeOffer offer = TradeOffer.builder()
                        .withFirstItem(new ItemStack(Material.DIAMOND, 1))
                        .withResultingItem(new ItemStack(Material.STICK, 1))
                        .build();
TradeOfferMeta meta = offer.getTradeMeta();
meta.setFirstItem(new ItemStack(Material.GOLD_INGOT, 1));
offer.setTradeMeta(meta);
```

I'm guessing it's doable, however, I can foresee issues with a TradeOffer being inherently mutable unless the method #setTradeMeta returns a clone of TradeOffer with the new meta (unlike ItemStack).

Wolvereness commented:
I don't see a purpose in having a 'meta' or having 'setters' anywhere.

gabizou commented:
> I don't see a purpose in having a 'meta' or having 'setters' anywhere.

So, there would be no link whatsoever from a TradeOffer to MerchantRecipe aside from constructing a TradeOffer from a MerchantRecipe for event throwing and org.bukkit.entity.Merchant#getOffers()?

Wolvereness commented:
> So, there would be no link ...

If you can remove almost every possible way to change data, other than a place that explicitly defines what is being changed, that's a good thing.

If an event is designed to change the 'result', then the only place to change that result should be in a method for that event called `setResult(...)`. If an event is designed to only be observed, based on the predicate state, then no part of that event should have any way to change the data, unless it's by virtue of the representation (like, providing a `location.clone()`). If we're adding something new to represent the data, why not make it immutable?

gabizou commented:
> If we're adding something new to represent the data, why not make it immutable?

Make sense. I'm just worried that opening the ability to set the TradeType.VANILLA to the API may be a bad thing.

Wolvereness commented:
If we need to _restrict_ a specific way to represent data, then it must be in CraftBukkit, with interfaces provided in Bukkit. If you think we need a way to differentiate trades, then it should be provided in the event. A `MultiMap<TradeData, TradeSource>` would work for that perfectly (as it stores a count, source, and data).

gabizou commented:
> A MultiMap<TradeData, TradeSource> would work for that perfectly (as it stores a count, source, and data).

I'm don't understand how that would work. It's probably because it's late and I should go to sleep. I'll read up on MultiMap and see how to implement it for this.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,468 @@
+        hash = hash * PRIME + resultingOffer.hashCode();
+        hash = hash * PRIME + uses;
+        hash = hash * PRIME + maxUses;
+        hash = hash * PRIME + (type == TradeType.VANILLA ? VANILLA : PLUGIN_ADDED);
```
```
type.hashCode()
```

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,468 @@
+        final int PRIME = 31, VANILLA = 1231, PLUGIN_ADDED = 1237;
+        int hash = 1;
+        hash = hash * PRIME + firstItem.hashCode();
+        hash = secondItem == null ? hash * PRIME : hash * PRIME + secondItem.hashCode();
```
```
hash = hash * PRIME + (secondItem == null ? 0 : secondItem.hashCode());
```

Wolvereness commented:
You haven't tested the serialization methods at all. You should make unit tests for that.

In addition, you leak `ItemStack` references in it.

Wolvereness commented:
I don't like the existence of many of those methods that return copies. Just add a method in the builder to copy values.

Considering it's in the same source / package, you can even use the fields as a shared secret (assuming you never leak it).

gabizou commented:
> You haven't tested the serialization methods at all. You should make unit tests for that.

Added tests to the CB PR.

> In addition, you leak ItemStack references in it.

Fixed.

> I don't like the existence of many of those methods that return copies. Just add a method in the builder to copy values.
> 
> Considering it's in the same source / package, you can even use the fields as a shared secret (assuming you never leak it).

Added.

amaranth commented on src/main/java/org/bukkit/event/merchant/MerchantRemoveOfferEvent.java:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * This event is called whenever a {@link org.bukkit.entity.Merchant}'s
+ * offers are being added to, whether it is a native Minecraft mechanic or
```
added to?

amaranth commented:
- Why do we care if the trade is from vanilla or a plugin? That doesn't seem like useful information.
- Why do you let someone modify the trade being removed? That could make the removal fail and if they want to remove something else they can cancel this event and do it with other API.
- You seem to have also designed this as a way to open custom trading GUIs based on your setTradingPlayer method and the fact that you can modify the trades shown without modifying the trades of the villager. If you want custom trading GUIs you should add support for them to openInventory and friends. I don't think these are appropriate for the villager API.
- Do the changes in 1.8 break the functionality of this plugin? They seem to break some adventure maps and a problem with an API like this in the past has been that Mojang keeps redesigning how trading works.

gabizou commented:
> Why do we care if the trade is from vanilla or a plugin? That doesn't seem like useful information.

I considered it useful in filtering out Vanilla trades in a MerchantTradeEvent such that a plugin may add a single new TradeOffer, while another plugin may also add custom TradeOffers and the end result would be a neat list of TradeOffers that are purely plugin generated. All of this without modifying the trades offered by the Merchant itself, reducing overhead and possibly excessive trades persisting from serialization and deserialization.

> Why do you let someone modify the trade being removed? That could make the removal fail and if they want to remove something else they can cancel this event and do it with other API.

True, the event was initially written when TradeOffers were mutable and I considered it allowable to alter the TradeOffer being removed. 

> You seem to have also designed this as a way to open custom trading GUIs based on your setTradingPlayer method and the fact that you can modify the trades shown without modifying the trades of the villager. If you want custom trading GUIs you should add support for them to openInventory and friends. I don't think these are appropriate for the villager API.

Correct, it isn't intended to behave solely for Villagers, as much as the NMS interface doesn't specify that the Merchant is a Villager, the internals show intention that it is supposed to be treated as an interface. This added API only adds that simpler functionality with room for modifications.

The setTradingPlayer method intended use is simply to activate a trade window forcibly between a player and a Merchant, in this case, a Villager. So I do see the need to have that method in the API, otherwise, this is reduced to listening for trade events and fiddling with internal NMS to activate said trade window. It is relative to removing the ability for Player.openWorkbench() because it is a customized workbench that may not exist. 

> Do the changes in 1.8 break the functionality of this plugin? They seem to break some adventure maps and a problem with an API like this in the past has been that Mojang keeps redesigning how trading works.

I haven't been able to read the fernflower decompiled source of any of the 1.8 snapshots specifically to see how MerchantRecipes are handled. Though I might add that come 1.8, if the IMerchant interface hasn't changed and then MerchantRecipe will add a new field/getter for experience, it is easily feasable to add to the TradeOffer API for this. The implementation may change as necessary, but the API would still function as intended as far as I understand. Unless Mojang started to send new update packets for dynamic Trade windows. 

It is true, being that 1.8 brings changes, we would have to see if the proposed API would need to change to add or remove functionality. However, it should still be understood that the TradeOffer will need changes for added "Tiers" and new methods may be necessary to force a specific trade tier on the Merchant. It is understandable that merchant and NPC may both need changing as Merchants may become the new Profession holder over NPC.

amaranth commented:
Player.openWorkbench is an example of what I'm saying you should make _instead of_ setTradingPlayer and the ability to modify the trades shown to the player without changing the trades in the villager. You would need something like Player.openTrading(Villager, TradeOffer[], boolean) with the boolean being to force opening it even if the villager doesn't exist or is too far away. If the array of offers is null you just use the trades from the villager. If the villager is null and so is the array IllegalArgumentException.

gabizou commented:
@amaranth 
So, I've spent some time to read the decompiled latest snapshot for how Mojang did indeed change IMerchant and EntityVillager. A few things to note that do make a deciding factor on this API:
- IMerchant is no longer implemented by one class, but now two. EntityVillager isn't the only class that implements IMerchant. This is why I was adamant about using Merchant for any references instead of Villager seeing as a Merchant will not always be a Villager (new class is a barebones implementation, but still a valid IMerchant).
- IMerchant does indeed have new methods and likewise EntityVillager was moderately refactored that would render the implementation of throwing a MerchantAddOfferEvent originating from the EntityVillager become useless. I still have found the core of where MerchantRecipe is being added to the Villager's recipe list, now that it's delegated through an interface, of which 4 new classes would have to be modified, so yes, this API functionality would technically be needing work come 1.8, but I'd be happy to help with updating the implementation.
- TradeType doesn't need to stay, I just thought it would be a nice thing to offer for plugin developers to understand where the TradeOffer originated from, but as Wolvereness explained, a TradeSource can be created specifically for this.
- Player#openTrade would most likely break entirely if implemented now in 1.7 and then re-implemented in 1.8 with the changes, since a ChatModule would have to be provided, along with the profession, and other information by NMS. It's simpler to spawn a villager and set it to trade with the Player instead of vice versa.

I am still remapping some obfuscated classes and methods to figure out what is going on with the MerchantRecipeList generation, though it seems that the new MerchantRecipes are pre-generated and added to a multi map that is queried for the various tiers depending on profession.

gabizou commented:
The following is a decompiled source of IMerchant with my own remappings of the methods:
https://gist.github.com/gabizou/25b4cc57375ca569930a

While this is the decompiled EntityVillager from the latest snapshot:
https://gist.github.com/gabizou/868c7006adc833a9f182

And this is the decompiled EntityPlayer#openTrade(IMerchant) method from the snapshot:
https://gist.github.com/gabizou/964c053442839dba90d4

Some of the classes are still obfuscated, but I've been doing my best to understand the important classes and figure out what they should be named based off the current mc-dev for 1.7

Granted, all of this may change up to the release of 1.8, but I believe it safe to say that leaving this with a proper Merchant interface that has a proper implementation will be better off than just allowing a Player#openTrade(Villager, TradeOffer[], boolean forced) to be implemented seeing as internally, it may have more issues.

gabizou commented:
Considering the option of HumanEntity#openTrade(Merchant merchant, TradeOffer[] offersToUse, boolean forced)
I've come up with a hasty implementation based on the serialization being performed on the EntityPlayer#openTrade(IMerchant, String)

https://gist.github.com/gabizou/3ad6c261c6a36b53b45f 

I found it semi appropriate for what is being handled with this type of trade window.

gabizou commented:
@amaranth As spoken about in IRC, the API additions and implementation in CB is overwritten now to the state that Merchant no longer has a link to the TradeOffers being offered, but a MerchantInventory that itself has the offers and can be modified. I've also created a dummy CraftDummyMerchant specifically for handling the custom TradeOffer list provided by HumanEntity#openTrade(TradeOffer[]).

If there's anything else that needs changing, let me know.

For the added features that were removed since our discussion, I'll ponder on how to re-implement them with this API and consider whether it necessary to add the API, and if so, I'll open another PR in the future if this PR is accepted.

amaranth commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +92,30 @@
+     * {@link org.bukkit.inventory.MerchantInventory#getOffers()}
+     *
+     * @param merchant The merchant to trade with. The merchant can not be null.
+     * @param force If false, and there is no Merchant at the location,
```
This feels like a copy/paste from openWorkbench but doesn't quite describe what is happening here. A null Merchant should be an IllegalArgumentException (and be documented as such) and the force flag is to override the vanilla distance check so the inventory stays open.

riking commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+     * @return true if this offer has nothing as a second item requirement
+     */
+    public boolean hasSecondItem() {
+        return this.secondItem != null && this.secondItem.getType() != Material.AIR;
```
But the type can never be AIR, right? The builder prevents that.

gabizou replied:
True. The builder does assign it as null if the second item is null.

riking commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+     */
+    public static ConfigurationSerializable deserialize(Map<String, Object> map) {
+        return builder()
+                .withFirstItem((ItemStack.deserialize((Map<String, Object>) map.get(FIRST_ITEM))))
```
This is wrong - `map.get(FIRST_ITEM)` should return an ItemStack, based on the behavior of other ConfigurationSerializables.

With the current code, this will CCE:

``` java
TradeOffer offer = ...;
TradeOffer serialCopy = TradeOffer.deserialize(offer.serialize());
```

gabizou replied:
But how are the serialization/deserialization tests passing then? 
The tests are performing these operations exactly and no exceptions are called.

riking replied:
Try this code.

``` java
        ItemStack item = new ItemStack(Material.LEATHER_BOOTS);
        LeatherArmorMeta meta = (LeatherArmorMeta) item.getItemMeta();
        meta.setColor(Color.PURPLE);
        item.setItemMeta(meta);
        TradeOffer offer = TradeOffer.builder()
                .withFirstItem(item)
                .withResultingItem(new ItemStack(Material.COAL))
                .build();
        config.set("TradeOffer", offer);

        TradeOffer.deserialize(offer.serialize());
```

```
java.lang.IllegalArgumentException: Cannot copy null stack
    at org.apache.commons.lang.Validate.notNull(Validate.java:203) ~[craftbukkit-1.7.9-R0.3-SNAPSHOT.jar:git-Bukkit-1.7.9-R0.1-41-g80bd834]
    at org.bukkit.inventory.ItemStack.<init>(ItemStack.java:127) ~[craftbukkit-1.7.9-R0.3-SNAPSHOT.jar:git-Bukkit-1.7.9-R0.1-41-g80bd834]
    at org.bukkit.inventory.TradeOffer.serialize(TradeOffer.java:268) ~[craftbukkit-1.7.9-R0.3-SNAPSHOT.jar:git-Bukkit-1.7.9-R0.1-41-g80bd834]
    at test.TestPlugin.putExamples(TestPlugin.java:92) ~[?:?]
    at test.TestPlugin.onEnable(TestPlugin.java:45) ~[?:?]
```

riking replied:
After applying a patch for serialize() to not try to provide a SECOND_ITEM value if it's null, we get this:

```
[19:31:24 INFO]: [TestPlugin] Enabling TestPlugin v1.0
[19:31:24 ERROR]: Error occurred while enabling TestPlugin v1.0 (Is it up to date?)
java.lang.ClassCastException: org.bukkit.inventory.ItemStack cannot be cast to java.util.Map
    at org.bukkit.inventory.TradeOffer.deserialize(TradeOffer.java:257) ~[craftbukkit-1.7.9-R0.3-SNAPSHOT.jar:git-Bukkit-1.7.9-R0.1-41-g80bd834]
    at test.TestPlugin.putExamples(TestPlugin.java:92) ~[?:?]
    at test.TestPlugin.onEnable(TestPlugin.java:45) ~[?:?]
    at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:316) ~[craftbukkit-1.7.9-R0.3-SNAPSHOT.jar:git-Bukkit-1.7.9-R0.1-41-g80bd834]
    at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:324) [craftbukkit-1.7.9-R0.3-SNAPSHOT.jar:git-Bukkit-1.7.9-R0.1-41-g80bd834]
```

riking replied:
Made a fix for you.

https://github.com/gabizou/Bukkit/pull/1/files

feildmaster commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,307 @@
+         */
+        public Builder copyOf(TradeOffer offer) {
+            Validate.notNull(offer, "Cannot copy from a null TradeType!");
+            return builder()
```
If you're going to force a new instance... you should just make this method static.

feildmaster commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,307 @@
+        public TradeOffer build() {
+            Validate.notNull(firstItem, "Cannot have a null first item in a TradeOffer!");
+            Validate.notNull(resultingOffer, "Cannot have a null resulting offer in a TradeOffer!");
+            return new TradeOffer(
```
logic would tell me to pass the Builder, and have the class worry about separating it... but I digress.

gabizou replied:
I've been thinking about this, and agreed, it makes for easier updating to TradeOffer in the future (such as experience granting)

Wolvereness commented:
Would have made a line comment, but the diff has changed:

> But how are the serialization/deserialization tests passing then? 

Because you haven't added:

```
assertThat(deserialized, is(original));
```

Speaking of which, were are the unit tests?

gabizou commented:
> Speaking of which, were are the unit tests?

In the CraftBukkit tests, I figured they would have to go there since I saw all of the ItemStack tests and ConfigurationSerialization tests being done there. 

> assertThat(deserialized, is(original));

That _is_ what I wrote in the CB test for deserialization of TradeOffer.

I'll also spend time this afternoon to re-read and clarify the documentation added for TradeOffer, MerchantInventory and Merchant as I'm sure I didn't update some documentation in certain places that elude to TradeOffer being mutable (from the initial writeup).

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+         * @throws IllegalArgumentException if the itemstack is null
+         * @throws IllegalArgumentException if the itemstack type is AIR
+         */
+        public Builder withFirstItem(ItemStack itemStack) throws IllegalArgumentException {
```
This method fails to treat the incoming ItemStack as malicious. You should not verify the material until after a safe copy has been made, with the copy under verification. Then, after validation, the field is to be set.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+         * @return This object, for chaining
+         * @throws IllegalArgumentException if the itemstack type is AIR
+         */
+        public Builder withSecondItem(ItemStack itemStack) throws IllegalArgumentException {
```
Two things.
1. I'd rather see `withoutSecondItem()` than `withSecondItem(null)`, but I do not recommend changing it without further feedback.
2. This method fails to treat the incoming ItemStack as malicious. You should not verify the material until after a safe copy has been made, with the copy under verification. Then, after validation, the field is to be set.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+         * @throws IllegalArgumentException if the itemstack is null
+         * @throws IllegalArgumentException if the itemstack type is AIR
+         */
+        public Builder withResultingItem(ItemStack itemStack) throws IllegalArgumentException {
```
This method fails to treat the incoming ItemStack as malicious. You should not verify the material until after a safe copy has been made, with the copy under verification. Then, after validation, the field is to be set.

gabizou commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+         */
+        public Builder withFirstItem(ItemStack itemStack) throws IllegalArgumentException {
+            Validate.notNull(itemStack, "Cannot have a null item for a TradeOffer's first item!");
+            Validate.isTrue(itemStack.getType() != Material.AIR, "Cannot have an AIR item for a TradeOffer's first item!");
```
@Wolvereness so, I should rewrite this to be:
Validate.isTrue(new ItemStack(itemStack).getType() != Material.AIR, "...");?

Wolvereness replied:
No.

Make the copy, store it as a local variable, verify it, then set the field.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+         * @param offer to use as a template
+         * @return This object, for chaining
+         */
+        public static Builder copyOf(TradeOffer offer) {
```
This method shouldn't be located in the builder - it should be an overloaded version of the method that creates a new builder (in parent class). Also, the 'offer' is trusted - we don't need to make defensive copies of the item stacks.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+     *
+     * @return a new copy of this offer
+     */
+    public TradeOffer copyOf() {
```
You don't need a method to copy something immutable.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+    private final int maxUses;
+    private final int uses;
+
+    TradeOffer(Builder builder) {
```
Builder has already made defensive copies - you don't need to copy them.

gabizou replied:
Fixed.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+         *
+         * @return The representive TradeOffer
+         */
+        public TradeOffer build() {
```
Build has no arguments, therefor, it should never throw `IllegalArgumentException`. `IllegalStateException` is the appropriate exception.

gabizou replied:
According to my IDE, Validate.validState(boolean, String) isn't available, is a simple if then throws ok?

amaranth replied:
Yeah, just doing the check manually works here.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+     *
+     * @return the second input item (or right item)
+     */
+    public ItemStack getSecondItem() {
```
All things considered, I think throwing an exception on null is more appropriate, with a `hasSecondItem()` to check.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Override
+    public Map<String, Object> serialize() {
```
These are not defensive copies, breaking the immutability of the entire class.

gabizou replied:
With the builder already making defensive copies, I didn't see the need to make defensive copies of the ItemStack being deserialized from the Map.

Wolvereness replied:
`((ItemStack) tradeOffer.serialize().get("first-item")).setMaterial(Material.AIR);`

gabizou replied:
Am I right in thinking that serializing an ItemStack will still link to the original ItemStack then?

gabizou replied:
Either way, fixed.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
This should not use the getters, which copy objects, and instead use the fields.

gabizou replied:
Fixed.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,301 @@
+         * @throws IllegalArgumentException if the itemstack is null
+         * @throws IllegalArgumentException if the itemstack type is AIR
+         */
+        public Builder withFirstItem(ItemStack itemStack) {
```
https://github.com/Bukkit/Bukkit/pull/1077#discussion_r14251356

> No.
> 
> Make the copy, store it as a local variable, verify it, then set the field.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,301 @@
+         * @return This object, for chaining
+         * @throws IllegalArgumentException if the itemstack type is AIR
+         */
+        public Builder withSecondItem(ItemStack itemStack) {
```
https://github.com/Bukkit/Bukkit/pull/1077#discussion_r14251356

> No.
> 
> Make the copy, store it as a local variable, verify it, then set the field.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,301 @@
+         * @throws IllegalArgumentException if the itemstack is null
+         * @throws IllegalArgumentException if the itemstack type is AIR
+         */
+        public Builder withResultingItem(ItemStack itemStack) {
```
https://github.com/Bukkit/Bukkit/pull/1077#discussion_r14251356

> No.
> 
> Make the copy, store it as a local variable, verify it, then set the field.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,301 @@
+     * @return A new Builder with the TradeOffer as a template for default
+     *     values
+     */
+    public Builder builderOf() {
```
This method should skip the API and just assign the fields directly, as the item stacks don't need to be copied.

Wolvereness commented:
Everywhere you return an `ItemStack`, you should be using `.clone()` instead of `new ItemStack(itemStack)`. The spots that you are receiving an `ItemStack` should keep the safe-copy behavior.

gabizou commented:
> When accepting an ItemStack via API that goes back into CB, always use new ItemStack(ItemStack), as we can't confirm the plugin doesn't have a malicious clone() method. ItemMeta instances are fine as long as they have been passed through the ItemFactory at some point (for example, in book events, it's compared to null, which throws an exception for malicious input).

From your first quote, I understood that the ItemStack should always be recreated by Bukkit's ItemStack. 

> Everywhere you return an ItemStack, you should be using .clone() instead of new ItemStack(itemStack)

From your second quote, am I to understand that the getters for TradeOffer should just return itemstack.clone(), but the builder setters should still perform new ItemStack(itemStack)?

feildmaster commented:
> From your second quote, am I to understand that the getters for TradeOffer should just return itemstack.clone(), but the builder setters should still perform new ItemStack(itemStack)?

Yes.

gabizou commented:
> From your second quote, am I to understand that the getters for TradeOffer should just return itemstack.clone(), but the builder setters should still perform new ItemStack(itemStack)?
> Yes.

Now to further expand on that, should the serialize() use ItemStack.clone() still or new ItemStack()?

feildmaster commented:
> Now to further expand on that, should the serialize() use ItemStack.clone() still or new ItemStack()?

And...

> Everywhere you return an ItemStack, you should be using .clone() instead of new ItemStack(itemStack)

Wolvereness commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,33 @@
 
     /**
+     * Opens a trade inventory window with the player's inventory on the bottom
+     * and the respective offers from the
```
This is wrapped incorrectly.

gabizou replied:
I'm confused, how is this wrapped incorrectly? Spacing? Linked class and method ignoring the 80 character limit?

Wolvereness commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,33 @@
+     * and the respective offers from the
+     * {@link org.bukkit.inventory.MerchantInventory#getOffers()}
+     *
+     * @param merchant The merchant to trade with. The merchant can not be null.
```
This is wrapped incorrectly.

Wolvereness commented on src/main/java/org/bukkit/inventory/MerchantInventory.java:
```diff
@@ -1,4 +1,46 @@
+
+/**
+ * Interface to the inventory of a Merchant. MerchantInventory is the only
+ * place to contain {@link org.bukkit.inventory.TradeOffer} list that a
```
This is wrapped incorrectly.

Wolvereness commented on src/main/java/org/bukkit/inventory/MerchantInventory.java:
```diff
@@ -1,4 +1,46 @@
+/**
+ * Interface to the inventory of a Merchant. MerchantInventory is the only
+ * place to contain {@link org.bukkit.inventory.TradeOffer} list that a
+ * Player is sent for
```
This is wrapped incorrectly.

Wolvereness commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,304 @@
+import org.bukkit.configuration.serialization.ConfigurationSerializable;
+
+/**
+ * Represents a Merchant's trade offer that always will have the following:
```
This is wrapped incorrectly.

I would also revise it to instead use a list.

Wolvereness commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,33 @@
+     * {@link org.bukkit.inventory.MerchantInventory#getOffers()}
+     *
+     * @param merchant The merchant to trade with. The merchant can not be null.
+     * @param force If false, and the merchant is too far away from the player,
```
This is wrapped incorrectly.

Wolvereness commented:
Eh, I see a lot of wrapping issues with the Javadoc, so I'm going to stop posting on those.

78 characters, the 79th is a \n

The notes following things like `@param` or `@return` are not complete sentences, so they wouldn't end in periods nor start with capitals.

gabizou committed:
Add Merchant Trade API - Adds BUKKIT-5663

gabizou commented:
> The notes following things like @param or @return are not complete sentences, so they wouldn't end in periods nor start with capitals.

And yet, I see in HumanEntity where the following is true:

```
    /**
     * Returns the ItemStack currently in your hand, can be empty.
     *
     * @return The ItemStack of the item you are currently holding.
     */
    public ItemStack getItemInHand();
```

> 78 characters, the 79th is a \n

So, is that to say that I should have my lines wrapping at 79 characters and the 80th as a \n?

> Eh, I see a lot of wrapping issues with the Javadoc,

I still don't understand what these wrapping issues are, I didn't find much documentation on the javadoc style standard for Bukkit/CraftBukkit aside from the pre-existing javadocs in Bukkit.

turt2live commented:
RE: Your concerns about wrapping:

Correct:

```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaa...
```

Incorrect:

```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaa...
```

gabizou commented:
@turt2live @Wolvereness I think I caught all the incorrectly wrapped javadocs. If not, I'll need some more pointers as to how they should be wrapped with an example case.

ST-DDT commented:
Nice API

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
 
     /**
+     * Opens a trade inventory window with the player's inventory on the bottom
+     * and the respective offers from the {@link org.bukkit.inventory.MerchantInventory#getOffers()}
```
You do not need to fully qualify this

Edit: and sentences end with periods.

gabizou replied:
> You do not need to fully qualify this

If the linked MerchantInventory isn't imported, should I import it manually for the sake of Javadocs?

turt2live replied:
You shouldn't need to import it...

gabizou replied:
> You shouldn't need to import it...

According to IntelliJ, I would, since it's not in the class imports.

ST-DDT replied:
Eclipse says so too.

turt2live replied:
That's IntelliJ, not JavaDocs.

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
+     *
+     * @param merchant the merchant to trade with
+     * @param force if false, and the merchant is too far away from the player,
+     *     the inventory will be closed if this player is too far
```
`the` should be on the line above

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
+     * @param force if false, and the merchant is too far away from the player,
+     *     the inventory will be closed if this player is too far
+     * @return the newly opened inventory view, or null if it could not be
+     *     opened
```
`opened` should be on the line above

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
+
+    /**
+     * Opens a trade inventory window with the player's inventory on the bottom
+     * and the specified offers listed in said trade inventory window.
```
`and` should be on the line above

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
+     * @param offers the offers to send to the player
+     * @param customName The custom name to label the trade window
+     * @return the newly opened inventory view, or null if it could not be
+     *     opened
```
`opened` should be n the line above

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
+     * @param customName The custom name to label the trade window
+     * @return the newly opened inventory view, or null if it could not be
+     *     opened
+     * @throws IllegalArgumentException if the offers are null
```
This is not required.

turt2live commented on src/main/java/org/bukkit/entity/HumanEntity.java:
```diff
@@ -90,6 +94,31 @@
+     *     the inventory will be closed if this player is too far
+     * @return the newly opened inventory view, or null if it could not be
+     *     opened
+     * @throws IllegalArgumentException if the merchant is null
```
This is not required.

turt2live commented on src/main/java/org/bukkit/event/merchant/MerchantAddOfferEvent.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Gets a copy of the list of offers being added to.
```
This is kinda strange to read. It feels somewhat broken without reading a bit more context.

ST-DDT replied:
I would use something like this:
`Gets a copy of the offers currently available for the merchant connected to this event.`
And maybe add something like this:
`Modifying the list or its entries does not have any impact on the merchant's offers.`

turt2live commented on src/main/java/org/bukkit/event/merchant/MerchantAddOfferEvent.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Gets the offer to add to the Merchant's list of offers.
```
This is also kinda strange

ST-DDT replied:
I would use something like this:
`Gets the offer that should be added to the merchants available offers.`

turt2live replied:
Still needs to define if it's mutable or not

turt2live commented on src/main/java/org/bukkit/event/merchant/MerchantEvent.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    /**
+     * Gets the merchant object itself.
```
`Gets the merchant involved with this event.`

turt2live commented on src/main/java/org/bukkit/event/merchant/MerchantEvent.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * Gets the merchant object itself.
+     *
+     * @return the merchant object
```
`The merchant`

turt2live commented on src/main/java/org/bukkit/inventory/MerchantInventory.java:
```diff
@@ -1,4 +1,44 @@
+ * Interface to the inventory of a Merchant. MerchantInventory is the only area
+ * to contain a {@link org.bukkit.inventory.TradeOffer} list that a Player is
+ * sent for {@link org.bukkit.entity.HumanEntity#openTrade(Merchant, boolean)}
+ */
```
You do not have to fully qualify the links. As well the wording is a bit off in general. I would start off by saying something like `Represents the inventory of a merchant` and go from there.

turt2live commented on src/main/java/org/bukkit/inventory/MerchantInventory.java:
```diff
@@ -1,4 +1,44 @@
 public interface MerchantInventory extends Inventory {
+
+    /**
+     * Gets the merchant belonging to this inventory
```
Period.

turt2live commented on src/main/java/org/bukkit/inventory/MerchantInventory.java:
```diff
@@ -1,4 +1,44 @@
+     * Adds the given TradeOffer to this Inventory's list of trade offers.
+     *
+     * @param offer to add
+     * @throws IllegalArgumentException if the offer is null
```
This is not required.

turt2live commented on src/main/java/org/bukkit/inventory/MerchantInventory.java:
```diff
@@ -1,4 +1,44 @@
+     * by this inventory.
+     *
+     * @param offer to remove
+     * @throws IllegalArgumentException if the offer is null
```
This is not required.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+
+/**
+ * Represents a Merchant's trade offer that always will have the following:
+ * <ul>
```
The list syntax we are moving forward with is as follows:

```
<ul>
    <li>
    ... content ...
    </li><li>
    ... content ...
    </li>
</ul>
```

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    /**
+     * This is a builder for TradeOffer.
```
These docs are not sufficient.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+        Builder() {}
+
+        /**
+         * Sets the first buying item for the trade offer.
```
Which is?

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         * @param itemStack the first buying item
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the itemstack is null
+         * @throws IllegalArgumentException if the itemstack type is AIR
```
`@throws` is not required.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         */
+        public Builder withFirstItem(ItemStack itemStack) {
+            Validate.notNull(itemStack, "Cannot have a null item for a TradeOffer's first item!");
+            ItemStack temp = new ItemStack(itemStack);
```
This is not how you should be cloning.

gabizou replied:
it's how @Wolvereness was telling me to clone, since ItemStack.clone() is not to be trusted with it being inherently able to be overridden.

turt2live replied:
Proof? I've been told otherwise many, many times.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+        public Builder withFirstItem(ItemStack itemStack) {
+            Validate.notNull(itemStack, "Cannot have a null item for a TradeOffer's first item!");
+            ItemStack temp = new ItemStack(itemStack);
+            Validate.isTrue(temp.getType() != Material.AIR, "Cannot have an AIR item for a TradeOffer's first item!");
```
Validate should be before the variable declaration

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+        }
+
+        /**
+         * Sets the second buying item for the trade offer.
```
Which is?

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         *
+         * @param itemStack the second buying item
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the itemstack type is AIR
```
Not required.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+                this.secondItem = null;
+                return this;
+            }
+            ItemStack temp = new ItemStack(itemStack);
```
Again, not how you should be cloning. Also move the validate to before the declaration.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         * @param itemStack the resulting offered item
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the itemstack is null
+         * @throws IllegalArgumentException if the itemstack type is AIR
```
`@throws` is not required

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         *
+         * @param uses the maximum uses this TradeOffer will have
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the uses is less than zero
```
Not required.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the uses is less than zero
+         */
+        public Builder withMaxUses(int uses) {
```
Will this still work with `Integer.MAX_VALUE`?

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         *
+         * @param uses the current uses this TradeOffer will have
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the uses is less than zero
```
Not required.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+         * @return this object, for chaining
+         * @throws IllegalArgumentException if the uses is less than zero
+         */
+        public Builder withSetUses(int uses) {
```
Can this be more than the maximum?

gabizou replied:
Technically, yes. The system accepts it.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+
+        /**
+         * Gets a new TradeOffer from the current state of this builder.
+         * <p>
```
`<p/>`

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    /**
+     * Gets the first item (or left item) required by this offer.
```
`left` does not really help us when we see a bunch of text in an editor.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+     * @return the second input item (or right item)
+     * @throws IllegalStateException if the item is null
+     */
+    public ItemStack getSecondItem() {
```
Why would the variable ever be null? And why does it have to throw an exception? Can't it just return null?

gabizou replied:
The variable would be null if there is no second item required for the TradeOffer. Removing the exceptions.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+    /**
+     * Returns true if this offer has a second item requirement.
+     *
+     * @return true if this offer has nothing as a second item requirement
```
This contradicts what you have previously documented.

turt2live commented on src/main/java/org/bukkit/inventory/TradeOffer.java:
```diff
@@ -0,0 +1,305 @@
+     *
+     * @return true if this offer's uses have exceeded the maximum uses
+     */
+    public boolean hasOfferExpired() {
```
Is the offer still attached to the merchant in any way?

gabizou replied:
Considering the TradeOffer is recreated each time when traversing between MinecraftServer <-> CraftBukkit, no. The method is supplied so that plugins will know of expired trade offers when they are queried, but by no means are the TradeOffers persistent/mutable.

turt2live commented on src/main/java/org/bukkit/merchant/Merchant.java:
```diff
@@ -0,0 +1,19 @@
+import org.bukkit.inventory.MerchantInventory;
+
+/**
+ * Represents a Merchant that can trade with a Player items for items
```
This is a bit weird to read.

gabizou commented:
> When accepting an ItemStack via API that goes back into CB, always use new ItemStack(ItemStack), as we can't confirm the plugin doesn't have a malicious clone() method. ItemMeta instances are fine as long as they have been passed through the ItemFactory at some point (for example, in book events, it's compared to null, which throws an exception for malicious input).

@turt2live As requested.

turt2live commented:
@gabizou source? The following is from this PR:

https://github.com/Bukkit/Bukkit/pull/1077#issuecomment-47265493

> Everywhere you return an ItemStack, you should be using .clone() instead of new ItemStack(itemStack). The spots that you are receiving an ItemStack should keep the safe-copy behavior.

gabizou commented:
> @gabizou source? The following is from this PR:

https://github.com/Bukkit/Bukkit/pull/1077/files#r15209292
In this case, the Builder is taking in an ItemStack, of which case, we must defensively copy as the comment mentions.
https://github.com/Bukkit/Bukkit/pull/1077#issuecomment-47265493

For all other cases where I am returning an ItemStack, I am performing ItemStack#clone() since it was already defensively copied from the Builder.

turt2live commented:
That still doesn't tell me where you got that quote from.

> I like pie. ~ Albert Einstein

gabizou commented:
> That still doesn't tell me where you got that quote from.

Ok, sorry, so we've linked to Wolvereness's comment twice now, I was doing as told, unless you're saying that I should go against what Wolvereness was saying about defensively copying the ItemStack as I have done in the builder.

For the sake of repetition:
https://github.com/Bukkit/Bukkit/pull/1077#issuecomment-46646563

Is it clear that I didn't know enough GitHub-comment-fu to link the direct comment itself?

ST-DDT commented:
The link does indeed state @Wolvereness option.
I guess he have to clearify his opinion himself here to solve your issues.

Maybe i'm wrong but i read the comment this way:
If you recieve the item always from a save source (Bukkit code) use `clone()`.
Use this version if you already made a defensive copy (in the constructor or somewhere else/Bukkit code) and should return that result as "immutable" copy.
If you potentially recieve an altered ItemStack from some plugin use the defensive cloning `new ItemStack(item)` to avoid issues due to corrupted clone methods in child classes.

Wolvereness commented:
@ST-DDT has it about right. I'd only make the exception for an _assumption_ that a certain code path is safe. If somewhere in the chain a plugin can implement an interface, or extend a class included in the Bukkit jar, then it's considered unsafe. Those places are gradually shrinking though, as more of the API becomes defensively safe.

gabizou committed:
Clean up javadocs.

gabizou commented:
@turt2live Updated the javadocs. I didn't change any of the `new ItemStack(ItemStack)` lines yet. I've removed most all of the `@throws Illegal....` as requested.
